### ISSUE TITLE
`undefined` validates against all schemas.

### ISSUE THREAD

**Issue by dugword** [tdegrunt/jsonschema#211]
I would expect the code below to throw, but `undefined` seems to be a valid object when validated against all schemas. When other falsy values are passed [null, false, NaN, '', 0] the validator correctly rejects them.

const assert = require('assert');
const jsonSchema = require('jsonschema');
const validator = new jsonSchema.Validator();

const validationResult = validator.validate(undefined, {
        "$schema": "http://json-schema.org/draft-04/schema#",
        "id": "http://foo.com/bar",
        "type": "object",
        "properties": { "foo": { "type": "string" } },
       "required": ["foo"]
        }, {
          propertyName: 'eventBatch',
        });

        assert.deepEqual(validationResult.errors, [], validationResult.errors.toString());

**Comment by awwright** [comment]
This is a tricky case because `undefined` isn't a JSON value of any kind. This package assumes that `{key: undefined}` means the same thing as not existing `{}`, so in the absence of `required: true`, it won't throw anything.

Perhaps the solution is to prohibit `undefined` as a value at all.

**Comment by Mike96Angelo** [comment]
The issue here is **not** when `instance = {key: undefined}` but when `instance = undefined`.

If `instance` is not a valid JSON type it should be invalid.

**Comment by awwright** [comment]
This behavior isn't intuitive—There should be some mechanism to prevent this from happening unintentionally, but it isn't wrong because of two things:

First, JSON Schema is permissive, it permits all values unless there's a rule making a value invalid.

Second, "undefined" is not well formed JSON, so how JSON Schema handles this is itself undefined.

Third, common usage in ECMAScript/JavaScript, the `undefined` value in a property typically has the same behavior as the property being missing altogether. So it doesn't make sense to make undefined a seventh type the same way `null` is.

The best workaround is just make sure that you're validating an object (`typeof instance==='object'` will ensure you're passing a JSON object, array, or null), or use `{required: true}` in your schema.

Is the suggestion that required: true should be the default for the top-level schema? Or is there another solution that might work?

**Comment by awwright** [comment]
@Mike96Angelo Actually maybe I misunderstood your point... JSON Schema doesn't make a distinction between `instance = null` and `instance = {key: null}` so in the same manner, it wouldn't make any distinction between `instance = undefined` and `instance = {key: undefined}`.

Is there a difference to you, if you used `null` instead?

**Comment by Mike96Angelo** [comment]
Yes there is a big difference and in my case I can't use `{ required: true }`.

Example:
```js
const schema = {
  type: 'object',
  required: [ 'a', 'b' ],
  properties: {
    a: { type: 'string' },
    b: { type: 'string' },
  },
}

jsonschema.validate(null, schema) // returns invalid

jsonschema.validate(undefined, schema) // returns valid
```

I understand that properties can be optional, but it seem wrong that the **root value** is optional.

Also where in the **JSON Schema Spec** is `{ required: true }` allowed, I can't seem to find it.  From what I understand a schema of `{ type: 'object' }` can specify an array of required properties (`{ required: [ 'property1', 'property2' ] }`).  I have not seen anything that states a schema of any type can have a field `{ required: true }`.

**Comment by awwright** [comment]
@Mike96Angelo "required" used to be boolean, and this form is supported as a historical artifact, from draft-03 I believe.

The validator works recursively, and doesn't make a distinction between properties and root values. Consider this:

```
jsonschema.validate({a:undefined}, {properties:{a:{type:"string"}}})
```

When it encounters the "a" property, it will recursively call jsonschema.validate:

```
jsonschema.validate(instance['a'], {type:"string"})
```

which is the same as `jsonschema.validate(undefined, {type:"string"})`. (It looks at all the "properties" schemas too, so it'll call that even if the property didn't exist at all.)

It's a quirk that doesn't really fit with how JSON Schema works these days, but it'll require a breaking change to fix.

The easiest way to work around this in the meantime will be to check that the root value being validated is a valid value in JSON. Is that very difficult for you to do for some reason?

**Comment by Mike96Angelo** [comment]
That seems fair, and the work arounds work fine.  I just found it non-intuitive (felt like a bug), but I guess its just how it was designed.  Thanks for taking the time to help me understand the problem.

**Comment by awwright** [comment]
This is a common enough issue that it's worth solving... In the next breaking change I'll add an option called `required`, which will default to true (whereas right now, the instance is not "required"). Set it to `false` for the v1 behavior.

Very few people should be impacted by this, but it's technically a breaking change so I'm scheduling it for the next major update, which should be 2.
